File: compute_exp_SSA.m
function data_cal=compute_exp_SSA(data,lambdaSC, lambdaAE )

%compute the pair-wise exponents and the SSA
%if there is several nephelometers with different wavelengths, lambda is a
%matrix with 4 columns corresponding to the 4 possibility (0, 1, 2 or
%without number)
%takes names
namesOK=data.Properties.VariableNames;
data_cal=timetable(data.Time);
CBs=startsWith(namesOK,'Bs');
namesBs=namesOK(CBs);
if ~isempty(namesBs)
    %take the green scat and the red abs
    a=contains(namesBs,'G');
    if sum(a)==0 && ~isempty(namesBs)
        a=1;
    end
    namesBsG=namesBs(a);

    CBbs=startsWith(namesOK,'Bbs');
    namesBbs=namesOK(CBbs);
    if ~isempty(namesBbs)
        a=contains(namesBbs,'G');
        if sum(a)==0 && ~isempty(namesBbs)
            a=1;
        end
        namesBbsG=namesBbs(a);

        %compute the backscattering fraction

        if length(namesBsG)==1 && length(namesBbsG)==1
            %namesBsG=namesBs(a);
            data_cal.BbsFG=data.(namesBbsG{1})./data.(namesBsG{1});
        else
            Cx(1,:)=contains(namesBsG,'0_')& ~contains(namesBsG,'dry') & ~contains(namesBsG,'Q');
            Cx(2,:)=contains(namesBsG,'1_')& ~contains(namesBsG,'dry') & ~contains(namesBsG,'Q');
            Cx(3,:)=contains(namesBsG,'dry');
            Cx(4,:)=contains(namesBsG,'_');
            Cx(4,:)=Cx(4,:) & ~Cx(1,:) & ~Cx(2,:) & ~Cx(3,:);

            Cy(1,:)=contains(namesBbsG,'0_')& ~contains(namesBbsG,'dry') & ~contains(namesBbsG,'Q');
            Cy(2,:)=contains(namesBbsG,'1_')& ~contains(namesBbsG,'dry') & ~contains(namesBbsG,'Q');
            Cy(3,:)=contains(namesBbsG,'dry');
            Cy(4,:)=contains(namesBbsG,'_');
            Cy(4,:)=Cy(4,:) & ~Cy(1,:) & ~Cy(2,:) & ~Cy(3,:);

            e=sum(Cx,2); r=sum(Cy,2);
            for i=1:length(e)
                if e(i)==1 & r(i)==1
                    namesBsx=namesBsG(Cx(i,:));
                    namesBbsx=namesBbsG(Cy(i,:));

                    if i==4
                        namesFG=strcat('BbsFG');
                    else
                        namesFG=strcat('BbsFG', num2str(i-1));
                    end
                    data_cal.(namesFG)=data.(namesBbsx{1})./data.(namesBsx{1});
                end
            end
        end
    else
        namesBsG=namesBs;
    end

    %compute the scattering exponent
    Cs(1,:)=contains(namesBs,'0_')& ~contains(namesBs,'dry') & ~contains(namesBs,'Q');
    Cs(2,:)=contains(namesBs,'1_')& ~contains(namesBs,'dry') & ~contains(namesBs,'Q');
    Cs(3,:)=contains(namesBs,'dry')& ~contains(namesBs,'Q');
    Cs(4,:)=contains(namesBs,'_')& ~contains(namesBs,'Q');
    Cs(4,:)=Cs(4,:) & ~Cs(1,:) & ~Cs(2,:) & ~Cs(3,:);
    for i=1:4
        namesBsx=namesBs(Cs(i,:));
        %ATTENTION: si il existe BsQ ne pas en tenir compte
        if length(namesBsx)>=3
            b=namesBsx(startsWith(namesBsx,'BsB'));
            g=namesBsx(startsWith(namesBsx,'BsG'));
            r=namesBsx(startsWith(namesBsx,'BsR'));
            N1=strcat('expS_bg', num2str(i-1));
            N2=strcat('expS_br', num2str(i-1));
            N3=strcat('expS_gr', num2str(i-1));
            data_cal.(N1)=real(-log(data.(b{1})./data.(g{1}))/log(lambdaSC(1)/lambdaSC(2)));
            data_cal.(N2)=real(-log(data.(b{1})./data.(r{1}))/log(lambdaSC(1)/lambdaSC(3)));
            data_cal.(N3)=real(-log(data.(g{1})./data.(r{1}))/log(lambdaSC(2)/lambdaSC(3)));
            %data_cal.exp_fit=Exp_SC_nan_regr(
        elseif length(namesBsx)==2
            b=namesBsx(startsWith(namesBsx,'BsB'));
            g=namesBsx(startsWith(namesBsx,'BsG'));
            N1=strcat('expS_bg', num2str(i-1));
            data_cal.(N1)=real(-log(data.(b{1})./data.(g{1}))/log(lambdaSC(1)/lambdaSC(2)));
        end
    end
else
    namesBsG=namesBs;
end

% absorption
CBa=startsWith(namesOK,'Ba');
namesBa=namesOK(CBa);
if ~isempty(namesBa)
    a=contains(namesBa,'R') |contains(namesBa,'5');
    if sum(a)==0 && ~isempty(namesBa)
        a=1;
    end
    namesBaR=namesBa(a);

    %compute the absorption exponent
    %ATTENTION: il peut y avoir AE + un a 3lambda
    ax=contains(namesBa,'Ba7') ;namesBa7=namesBa(ax);
    if  ~isempty(namesBa7) & length(namesBa)==7
        b=namesBa(startsWith(namesBa,'Ba2'));
        g=namesBa(startsWith(namesBa,'Ba3'));
        r=namesBa(startsWith(namesBa,'Ba5'));
        data_cal.expA_bg=real(-log(data.(b{1})./data.(g{1}))/log(470/520));
        data_cal.expA_br=real(-log(data.(b{1})./data.(r{1}))/log(470/660));
        data_cal.expA_gr=real(-log(data.(g{1})./data.(r{1}))/log(520/660));
        a1=namesBa(startsWith(namesBa,'Ba1'));
        a4=namesBa(startsWith(namesBa,'Ba4'));
        a6=namesBa(startsWith(namesBa,'Ba6'));
        a7=namesBa(startsWith(namesBa,'Ba7'));
        all_Abs=[data.(a1{1}) data.(b{1}) data.(g{1}) data.(a4{1}) data.(r{1}) data.(a6{1}) data.(a7{1})];
        data_cal.expA_fit=Exp_AE_nan_regr(all_Abs);
    end
    if  length(namesBa)==10 || length(namesBa)==13 || length(namesBa)==16
        warning(' one AE and one/several 3w instrument?');
        a1=namesBa(startsWith(namesBa,'Ba1'));
        b=namesBa(startsWith(namesBa,'Ba2'));
        g=namesBa(startsWith(namesBa,'Ba3'));
        a4=namesBa(startsWith(namesBa,'Ba4'));
        r=namesBa(startsWith(namesBa,'Ba5'));
        a6=namesBa(startsWith(namesBa,'Ba6'));
        a7=namesBa(startsWith(namesBa,'Ba7'));
        data_cal.expA_AE_bg=real(-log(data.(b{1})./data.(g{1}))/log(470/520));
        data_cal.expA_AE_br=real(-log(data.(b{1})./data.(r{1}))/log(470/660));
        data_cal.expA_AE_gr=real(-log(data.(g{1})./data.(r{1}))/log(520/660));
        all_Abs=[data.(a1{1}) data.(b{1}) data.(g{1}) data.(a4{1}) data.(r{1}) data.(a6{1}) data.(a7{1})];
        data_cal.expA_AE_fit=Exp_AE_nan_regr(all_Abs);

        Ca(1,:)=contains(namesBa,'0_');
        Ca(2,:)=contains(namesBa,'1_');
        Ca(3,:)=contains(namesBa,'dry');
        Ca(4,:)=contains(namesBa,'_');
        Ca(4,:)=Ca(4,:) & ~Ca(1,:) & ~Ca(2,:) & ~Ca(3,:);
        for i=1:4
            namesBax=namesBa(Ca(i,:));
            %ATTENTION: changer les noms sinon ils se récrivent l'un sur
            %l'autre
            if length(namesBax)==3
                ba=namesBax(startsWith(namesBax,'BaB'));
                ga=namesBax(startsWith(namesBax,'BaG'));
                ra=namesBax(startsWith(namesBax,'BaR'));
                N1=strcat('expA_bg', num2str(i-1));
                N2=strcat('expA_br', num2str(i-1));
                N3=strcat('expA_gr', num2str(i-1));
                data_cal.(N1)=real(-log(data.(ba{1})./data.(ga{1}))/log(lambdaAE(1)/lambdaAE(2)));
                data_cal.(N2)=real(-log(data.(ba{1})./data.(ra{1}))/log(lambdaAE(1)/lambdaAE(3)));
                data_cal.(N3)=real(-log(data.(ga{1})./data.(ra{1}))/log(lambdaAE(2)/lambdaAE(3)));
            end
        end
    elseif length(namesBa)==12
        %2 x 3w + 2 size cuts
        Ca(1,:)=contains(namesBa,'0_A11');
        Ca(2,:)=contains(namesBa,'1_A11');
        Ca(3,:)=contains(namesBa,'0_A12');
        Ca(4,:)=contains(namesBa,'1_A12');
        % Ca(4,:)=Ca(4,:) & ~Ca(1,:) & ~Ca(2,:) & ~Ca(3,:);
        for i=1:4
            namesBax=namesBa(Ca(i,:));
            %ATTENTION: changer les noms sinon ils se récrivent l'un sur
            %l'autre
            if length(namesBax)==3
                ba=namesBax(startsWith(namesBax,'BaB'));
                ga=namesBax(startsWith(namesBax,'BaG'));
                ra=namesBax(startsWith(namesBax,'BaR'));
                N1=strcat('expA_bg', num2str(i-1));
                N2=strcat('expA_br', num2str(i-1));
                N3=strcat('expA_gr', num2str(i-1));
                data_cal.(N1)=real(-log(data.(ba{1})./data.(ga{1}))/log(lambdaAE(1)/lambdaAE(2)));
                data_cal.(N2)=real(-log(data.(ba{1})./data.(ra{1}))/log(lambdaAE(1)/lambdaAE(3)));
                data_cal.(N3)=real(-log(data.(ga{1})./data.(ra{1}))/log(lambdaAE(2)/lambdaAE(3)));
            end
        end
    else
        Ca(1,:)=contains(namesBa,'0_');
        Ca(2,:)=contains(namesBa,'1_');
        Ca(3,:)=contains(namesBa,'dry');
        Ca(4,:)=contains(namesBa,'_');
        Ca(4,:)=Ca(4,:) & ~Ca(1,:) & ~Ca(2,:) & ~Ca(3,:);
        for i=1:4
            namesBax=namesBa(Ca(i,:));
            %ATTENTION: changer les noms sinon ils se récrivent l'un sur
            %l'autre
            if length(namesBax)==3
                ba=namesBax(startsWith(namesBax,'BaB'));
                ga=namesBax(startsWith(namesBax,'BaG'));
                ra=namesBax(startsWith(namesBax,'BaR'));
                N1=strcat('expA_bg', num2str(i-1));
                N2=strcat('expA_br', num2str(i-1));
                N3=strcat('expA_gr', num2str(i-1));
                data_cal.(N1)=real(-log(data.(ba{1})./data.(ga{1}))/log(lambdaAE(1)/lambdaAE(2)));
                data_cal.(N2)=real(-log(data.(ba{1})./data.(ra{1}))/log(lambdaAE(1)/lambdaAE(3)));
                data_cal.(N3)=real(-log(data.(ga{1})./data.(ra{1}))/log(lambdaAE(2)/lambdaAE(3)));
            end
        end
    end
end

if ~isempty(namesBsG) && ~isempty(namesBa)
    %compute the SSA with, if possible the green scat and the red abs
    %or with any wavelength, since only the trend is important
    CaR(1,:)=startsWith(namesBaR,'BaR0');
    CaR(2,:)=startsWith(namesBaR,'BaR1');
    CaR(3,:)=startsWith(namesBaR,["BacR";"Ba3"]) ;
    CaR(4,:)=startsWith(namesBaR,["BaR_";"Ba5"]);
    CaG=startsWith(namesBa,'BaG_');

    CsG(1,:)=startsWith(namesBsG,'BsG0');
    CsG(2,:)=startsWith(namesBsG,'BsG1');
    CsG(3,:)=startsWith(namesBsG,'BsG') & endsWith(namesBsG,'dry');
    CsG(4,:)=startsWith(namesBsG,'BsG_');
    for i=1:4
        if nansum(CaR(i,:))>0 && nansum(CsG(i,:))>0
            namesBsy=namesBsG(CsG(i,:));
            namesBay=namesBaR(CaR(i,:));
            for j=1:length(namesBsy)
                for k=1:length(namesBay)
                    N_SSA=strcat('SSA', num2str(i-1),num2str(j),num2str(k));
                    data_cal.(N_SSA)=data.(namesBsy{j})./(data.(namesBsy{j})+data.(namesBay{k}));
                end
            end

        elseif nansum(CaG)>0 && nansum(CsG(i,:))>0
            namesBsy=namesBsG(CsG(i,:));
            namesBay=namesBa(CaG);
            for j=1:length(namesBsy)
                for k=1:length(namesBay)
                    N_SSA=strcat('SSA', num2str(i-1),num2str(j),num2str(k));
                    data_cal.(N_SSA)=data.(namesBsy{j})./(data.(namesBsy{j})+data.(namesBay{k}));
                end
            end
        end
    end
end


end

File: compute_exp_SSA.py
import pandas as pd
import numpy as np

def Exp_AE_nan_regr(data):

    lams = np.array([370, 470, 520, 590, 660, 880, 950], dtype=float)
    x = np.log(lams)

    data = np.asarray(data, dtype=float)

    # Evitar logs de valors <= 0
    data[data <= 0] = np.nan

    y = np.log(data)

    aae = np.full(len(y), np.nan)

    for i in range(len(y)):
        mask = np.isfinite(y[i])

        # mínim 2 punts per fer regressió
        if np.sum(mask) >= 2:
            slope, intercept = np.polyfit(x[mask], y[i,mask], 1)
            aae[i] = -slope

    return aae

def compute_exp_ssa(data, lambdaSC, lambdaAE):
    """
    Parameters
    ----------
    data : pandas.DataFrame
        DataFrame amb columnes tipus:
        BsB, BsG, BsR, BbsB, BbsG, BbsR

    lambdaSC : list or array
        Longituds d'ona [blue, green, red]

    Returns
    -------
    data_cal : pandas.DataFrame
    """

    data_cal = pd.DataFrame(index=data.index)

    names = data.columns.tolist()

    # =========================================================
    # SCATTERING VARIABLES
    # =========================================================

    namesBs = [c for c in names if c.startswith("Bs")]

    if len(namesBs) > 0:

        # GREEN SCATTERING
        namesBsG = [c for c in namesBs if "G" in c]

        if len(namesBsG) == 0:
            namesBsG = [namesBs[0]]

        # BACKSCATTERING
        namesBbs = [c for c in names if c.startswith("Bbs")]

        if len(namesBbs) > 0:

            namesBbsG = [c for c in namesBbs if "G" in c]

            if len(namesBbsG) == 0:
                namesBbsG = [namesBbs[0]]

            # Backscattering fraction
            if len(namesBsG) == 1 and len(namesBbsG) == 1:

                data_cal["BbsFG"] = (data[namesBbsG[0]] / data[namesBsG[0]])
            
            else:

                Cx = {} # CLASSIFY BsG CHANNELS

                Cx[0] = [("0_" in c) and ("dry" not in c) and ("Q" not in c)
                    for c in namesBsG]

                Cx[1] = [("1_" in c) and ("dry" not in c) and ("Q" not in c)
                    for c in namesBsG]

                Cx[2] = [("dry" in c)
                    for c in namesBsG]

                Cx[3] = [("_" in c) and not Cx[0][i] and not Cx[1][i] and not Cx[2][i]
                    for i, c in enumerate(namesBsG)]

                Cy = {} # CLASSIFY BbsG CHANNELS

                Cy[0] = [("0_" in c) and ("dry" not in c) and ("Q" not in c)
                    for c in namesBbsG]

                Cy[1] = [("1_" in c) and ("dry" not in c) and ("Q" not in c)
                    for c in namesBbsG]

                Cy[2] = [("dry" in c)
                    for c in namesBbsG]

                Cy[3] = [("_" in c) and not Cx[0][i] and not Cx[1][i] and not Cx[2][i]
                    for i, c in enumerate(namesBbsG)]
                
                for i in range(4):

                    # how many channels found
                    e = sum(Cx[i])
                    r = sum(Cy[i])

                    # only if exactly one Bs and one Bbs
                    if e == 1 and r == 1:

                        # get matching names
                        namesBsx = [namesBsG[j] for j in range(len(namesBsG))
                            if Cx[i][j]]

                        namesBbsx = [namesBbsG[j] for j in range(len(namesBbsG))
                            if Cy[i][j]]

                        # output variable name
                        if i == 3:
                            namesFG = "BbsFG"
                        else:
                            namesFG = f"BbsFG{i}"

                        # compute backscattering fraction
                        data_cal[namesFG] = (data[namesBbsx[0]] / data[namesBsx[0]])
        
        namesBsG=namesBs

        # COMPUTE SCATTERING ANGSTROM EXPONENT

        # CLASSIFY Bs CHANNELS
        Cs = {} 

        Cs[0] = [("0_" in c) and ("dry" not in c) and ("Q" not in c)
            for c in namesBs]

        Cs[1] = [("1_" in c) and ("dry" not in c) and ("Q" not in c)
            for c in namesBs]

        Cs[2] = [("dry" in c) and ("Q" not in c)
            for c in namesBs]

        Cs[3] = [("_" in c) and ("Q" not in c) and not Cs[0][i] and not Cs[1][i] and not Cs[2][i]
            for i, c in enumerate(namesBs)]
        
        # LOOP OVER CONFIGURATIONS        
        for i in range(4): 

            namesBsx = [namesBs[j] for j in range(len(namesBs)) if Cs[i][j]]

            if len(namesBsx) >= 3: # THREE WAVELENGTHS

                b = [c for c in namesBsx if c.startswith("BsB")]
                g = [c for c in namesBsx if c.startswith("BsG")]
                r = [c for c in namesBsx if c.startswith("BsR")]
                
                if len(b) > 0 and len(g) > 0: # BLUE-GREEN

                    data_cal[f"expS_bg{i}"] = (-np.log(data[b[0]] / data[g[0]])
                        / np.log(lambdaSC[0] / lambdaSC[1]))

                if len(b) > 0 and len(r) > 0: # BLUE-RED

                    data_cal[f"expS_br{i}"] = (-np.log(data[b[0]] / data[r[0]])
                        / np.log(lambdaSC[0] / lambdaSC[2]))

                if len(g) > 0 and len(r) > 0: # GREEN-RED

                    data_cal[f"expS_gr{i}"] = (-np.log(data[g[0]] / data[r[0]])
                        / np.log(lambdaSC[1] / lambdaSC[2]))
            
            elif len(namesBsx) == 2: # ONLY TWO WAVELENGTHS

                b = [c for c in namesBsx if c.startswith("BsB")]
                g = [c for c in namesBsx if c.startswith("BsG")]

                if len(b) > 0 and len(g) > 0:

                    data_cal[f"expS_bg{i}"] = (-np.log(data[b[0]] / data[g[0]])
                        / np.log(lambdaSC[0] / lambdaSC[1]))
    
    else:
        namesBsG=namesBs

    # =========================================================
    # ABSORPTION VARIABLES
    # =========================================================

    namesBa = [c for c in names if c.startswith("Ba")]

    if len(namesBa) > 0:

        # FIND RED CHANNELS
        namesBaR = [c for c in namesBa if ("R" in c) or ("5" in c)]

        if len(namesBaR) == 0:
            namesBaR = [namesBa[0]]

        # AE33 / 7 WAVELENGTHS
        namesBa7 = [c for c in namesBa if "Ba7" in c]

        # CLASSIFY Bs CHANNELS
        Ca = {}
        Ca[0] = [("0_" in c) for c in namesBa]
        Ca[1] = [("1_" in c) for c in namesBa]
        Ca[2] = [("dry" in c) for c in namesBa]
        Ca[3] = [("_" in c) and not Ca[0][i] and not Ca[1][i] and not Ca[2][i]
            for i, c in enumerate(namesBa)]
        
        for i in range(4):
            namesBax = [namesBa[j] for j in range(len(namesBa))
                    if Ca[i][j]]

            if len(namesBa7) > 0 and len(namesBax) == 7:

                b = [c for c in namesBax if c.startswith("Ba2")]
                g = [c for c in namesBax if c.startswith("Ba3")]
                r = [c for c in namesBax if c.startswith("Ba5")]

                # AAE
                data_cal[f"expA_bg{i}"] = (-np.log(data[b[0]] / data[g[0]])
                    / np.log(470 / 520))

                data_cal[f"expA_br{i}"] = (-np.log(data[b[0]] / data[r[0]])
                    / np.log(470 / 660))

                data_cal[f"expA_gr{i}"] = (-np.log(data[g[0]] / data[r[0]])
                    / np.log(520 / 660))

                a1 = [c for c in namesBax if c.startswith("Ba1")]
                a4 = [c for c in namesBax if c.startswith("Ba4")]
                a6 = [c for c in namesBax if c.startswith("Ba6")]
                a7 = [c for c in namesBax if c.startswith("Ba7")]

                all_abs = np.column_stack([data[a1[0]],data[b[0]],data[g[0]],
                    data[a4[0]],data[r[0]],data[a6[0]],data[a7[0]]])

                data_cal[f"expA_fit{i}"] = Exp_AE_nan_regr(all_abs)

            # MIXED AE + 3W INSTRUMENTS
            elif len(namesBax) in [10, 13, 16]:

                print("Warning: one AE and one/several 3w instrument?")

                # AE33 PART
                b = [c for c in namesBax if c.startswith("Ba2")]
                g = [c for c in namesBax if c.startswith("Ba3")]
                r = [c for c in namesBax if c.startswith("Ba5")]
                a1 = [c for c in namesBax if c.startswith("Ba1")]
                a4 = [c for c in namesBax if c.startswith("Ba4")]
                a6 = [c for c in namesBax if c.startswith("Ba6")]
                a7 = [c for c in namesBax if c.startswith("Ba7")]

                data_cal[f"expA_AE_bg{i}"] = (-np.log(data[b[0]] / data[g[0]])
                    / np.log(470 / 520))

                data_cal[f"expA_AE_br{i}"] = (-np.log(data[b[0]] / data[r[0]])
                    / np.log(470 / 660))

                data_cal[f"expA_AE_gr{i}"] = (-np.log(data[g[0]] / data[r[0]])
                    / np.log(520 / 660))

                all_abs = np.column_stack([data[a1[0]],data[b[0]],data[g[0]],
                    data[a4[0]],data[r[0]],data[a6[0]],data[a7[0]]])
                data_cal[f"expA_fit{i}"] = Exp_AE_nan_regr(all_abs)

                # CLASSIFY 3-WAVELENGTH INSTRUMENTS
                ba = [c for c in namesBax if c.startswith("BaB")]
                ga = [c for c in namesBax if c.startswith("BaG")]
                ra = [c for c in namesBax if c.startswith("BaR")]

                data_cal[f"expA_bg{i}"] = (-np.log(data[ba[0]] / data[ga[0]])
                    / np.log(lambdaAE[0] / lambdaAE[1]))
                
                data_cal[f"expA_br{i}"] = (-np.log(data[ba[0]] / data[ra[0]])
                    / np.log(lambdaAE[0] / lambdaAE[2]))
                
                data_cal[f"expA_gr{i}"] = (-np.log(data[ga[0]] / data[ra[0]])
                    / np.log(lambdaAE[1] / lambdaAE[2]))

            # 2 instruments x 2 size cuts x 3 wavelengths
            elif len(namesBax) == 12:

                Ca = {}
                Ca[0] = [("0_A11" in c) for c in namesBax]
                Ca[1] = [("1_A11" in c) for c in namesBax]
                Ca[2] = [("0_A12" in c) for c in namesBax]
                Ca[3] = [("1_A12" in c) for c in namesBax]

                for i in range(4):
                
                    namesBax = [namesBa[j] for j in range(len(namesBa))
                        if Ca[i][j]]

                    if len(namesBax) == 3:
                    
                        ba = [c for c in namesBax if c.startswith("BaB")]
                        ga = [c for c in namesBax if c.startswith("BaG")]
                        ra = [c for c in namesBax if c.startswith("BaR")]

                        data_cal[f"expA_bg{i}"] = (-np.log(data[ba[0]] / data[ga[0]])
                            / np.log(lambdaAE[0] / lambdaAE[1]))

                        data_cal[f"expA_br{i}"] = (-np.log(data[ba[0]] / data[ra[0]])
                            / np.log(lambdaAE[0] / lambdaAE[2]))

                        data_cal[f"expA_gr{i}"] = (-np.log(data[ga[0]] / data[ra[0]])
                            / np.log(lambdaAE[1] / lambdaAE[2]))

            else:
                if len(namesBax) == 3:
                
                    ba = [c for c in namesBax if c.startswith("BaB")]
                    ga = [c for c in namesBax if c.startswith("BaG")]
                    ra = [c for c in namesBax if c.startswith("BaR")]
                    data_cal[f"expA_bg{i}"] = (-np.log(data[ba[0]] / data[ga[0]])
                        / np.log(lambdaAE[0] / lambdaAE[1]))
                    data_cal[f"expA_br{i}"] = (-np.log(data[ba[0]] / data[ra[0]])
                        / np.log(lambdaAE[0] / lambdaAE[2]))
                    data_cal[f"expA_gr{i}"] = (-np.log(data[ga[0]] / data[ra[0]])
                        / np.log(lambdaAE[1] / lambdaAE[2]))            

    # =========================================================
    # COMPUTE SSA
    # =========================================================

    if len(namesBsG) > 0 and len(namesBa) > 0:

        # CLASSIFY ABSORPTION
        CaR = {}
        CaR[0] = [c.startswith("BaR0") for c in namesBaR]
        CaR[1] = [c.startswith("BaR1") for c in namesBaR]
        CaR[2] = [(("BacR" in c) or ("Ba3" in c)) for c in namesBaR]
        CaR[3] = [(c.startswith("BaR_") or "Ba5" in c) for c in namesBaR]

        CaG = [c.startswith("BaG_") for c in namesBa]

        # CLASSIFY SCATTERING
        CsG = {}
        CsG[0] = [c.startswith("BsG0") for c in namesBsG]
        CsG[1] = [c.startswith("BsG1") for c in namesBsG]
        CsG[2] = [(c.startswith("BsG") and c.endswith("dry")) for c in namesBsG]
        CsG[3] = [c.startswith("BsG_") for c in namesBsG]
        
        # LOOP OVER GROUPS
        for i in range(4):

            # CASE 1: matching BaR + BsG
            if sum(CaR[i]) > 0 and sum(CsG[i]) > 0:

                namesBsy = [namesBsG[j] for j in range(len(namesBsG)) if CsG[i][j]]
                namesBay = [namesBaR[j] for j in range(len(namesBaR)) if CaR[i][j]]

                for j in namesBsy:
                    for k in namesBay:

                        N_SSA = f"SSA{i}{j}{k}"

                        data_cal[N_SSA] = (data[j] / (data[j] + data[k]))

            # CASE 2: fallback Ba
            elif sum(CaG) > 0 and sum(CsG[i]) > 0:

                namesBsy = [namesBsG[j] for j in range(len(namesBsG)) if CsG[i][j]]
                namesBay = [namesBa[j] for j in range(len(namesBa)) if CaG[j]]

                for j in namesBsy:
                    for k in namesBay:

                        name = f"SSA{i}{j}{k}"

                        data_cal[name] = (data[j] / (data[j] + data[k]))

    return data_cal